daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/DataValue.scala

```diff
@@ -0,0 +1,209 @@
+trait Nullable extends tUseNilForDefault;
+trait tUseNilForDefault;
+sealed trait DataValuePrimitiveType
+
+/*
+ * The type variable X here is a phantom type. That is to say, it is not actually
+ * used to describe the type of anything within DataValue. The goal is to provide us
+ * with more ability to control the inheritense relationship between various DataValue types.
+ * In theory, this would be accomplished by establishing a type hierarchy on the types
+ * used for T. Indeed, if Scala had type-classes as a first-class feature that is likely
+ * the approach we would have taken. However, since most of the types we use for T
+ * are out of our control, we are unable to control what type structure they have.
+ * By carefully selecting X in our type aliases, it is possible for to define types
+ * that can represent any subset of types which we want. It is further possible to
+ * allow for automatic upcasting between these types where appropriate.
+ *
+ * We inline all methods because attempts to call non-inlined methods trigger boxing
+ */
+final class DataValue[+T <: AnyRef, +X <: AnyRef] private (val v: T) extends AnyVal {
+  @inline def isEmpty = DataValue.NoValue.v eq v
+  @inline def isDefined = !isEmpty
+  @inline def value = v
+  @inline override def toString = if (isEmpty) "NoValue" else "DataValue(" + v.toString + ")"
+
+  @inline def getAnyRef = v.asInstanceOf[AnyRef]
+  @inline def getBigDecimal = v.asInstanceOf[JBigDecimal]
+  @inline def getCalendar = v.asInstanceOf[DFDLCalendar]
+  @inline def getDate = v.asInstanceOf[DFDLDate]
+  @inline def getTime = v.asInstanceOf[DFDLTime]
+  @inline def getDateTime = v.asInstanceOf[DFDLDateTime]
+  @inline def getByteArray = v.asInstanceOf[Array[Byte]]
+  @inline def getBoolean = v.asInstanceOf[JBoolean]
+  @inline def getNumber = v.asInstanceOf[JNumber]
+  @inline def getInt = v.asInstanceOf[JInt]
+  @inline def getLong = v.asInstanceOf[JLong]
+  @inline def getDouble = v.asInstanceOf[JDouble]
+  @inline def getBigInt = v.asInstanceOf[JBigInt]
+  @inline def getString = v.asInstanceOf[JString]
+  @inline def getURI = v.asInstanceOf[URI]
+
+  @inline def getNonNullable = this.asInstanceOf[DataValue[T, X with NonNullable]]
+  @inline def getNullablePrimitive = this.asInstanceOf[DataValue.DataValuePrimitiveNullable]
+
+  @inline def getOptionAnyRef = {
+    if (isEmpty) {
+      None
+    } else {
+      Some(getAnyRef)
+    }
+  }
+
+  @inline def getMaybe[T <: AnyRef]: Maybe[T] = {
+    if (isEmpty) {
+      Nope
+    } else {
+      One(getAnyRef.asInstanceOf[T])
+    }
+  }
+}
+
+object DataValue {
+  type DataValueAny = DataValue[AnyRef, Nullable]
+  type DataValueNonEmpty = DataValue[AnyRef, NonNullable]
+  type DataValuePrimitive = DataValue[AnyRef, NonNullable with DataValuePrimitiveType]
+  type DataValuePrimitiveNullable = DataValue[AnyRef, Nullable with DataValuePrimitiveType]
+  type DataValuePrimitiveOrUseNilForDefaultOrNull = DataValue[AnyRef, tUseNilForDefault]
```

Comment:
This one I think is trying to make a distinction we don't need. UseNilForDefault sentinel is just another notion of NoValue plus "oh yeah and if you're considering defaulting this thing, then here's another tidbit of info".

Reply:
Ah. In the above comment I hadn't realized that UseNilForDefault doesn't actually show up as the value of a DISimple. It only appears in the ERD to signify that the value, if not present, can be defaulted not to the default value, but to nil.  
So I withdraw the suggestion.